pkg/storage/stores/shipper/indexshipper/tsdb/compactor.go

```diff
@@ -205,6 +206,42 @@ func (t *tableCompactor) CompactTable() error {
 	return nil
 }
 
+// processSourceIndex processes a single source index file by downloading it,
+// reading its series, and adding them to the builder. The file is closed and removed
+// using defer statements to ensure cleanup happens even if errors occur.
+func processSourceIndex(ctx context.Context, sourceIndex storage.IndexFile, sourceIndexSet compactor.IndexSet, builder *Builder) error {
+	path, err := sourceIndexSet.GetSourceFile(sourceIndex)
+	if err != nil {
+		return err
+	}
+
+	indexFile, err := OpenShippableTSDB(path)
+	if err != nil {
+		// Clean up the downloaded file if we can't open it
+		if removeErr := os.Remove(path); removeErr != nil {
```

Comment:
Let's move this up in `defer` just after the nil error check from `GetSourceFile`. We can then remove the duplicate `os.Remove` from the other `defer` statement on line 228.